[Noetic] Implement pose graph node marginalization

hidmic opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | https://github.com/SteveMacenski/slam_toolbox/issues/407 (partial) |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Clearpath Jackal Gazebo simulation |

---

## Description of contribution in a few bullet points

* Added graph node marginalization support using the approximations proposed by Kretzschmar et. al. Had to fall back to generalized matrix inverses to cope with constant parameter blocks in the solver.
* Modified lifelong support to marginalize decayed nodes instead of plainly removing them.

## Description of documentation updates required from your changes

* (I think) I left enough comments to walk back from the implementation to the paper, but do take a look.

---

## Future work that may be required in bullet points

* I wonder if Karto wouldn't be better off fully embracing Eigen.

hidmic commented:
@kmilo7204 I still have to document the procedure I've used to test this and probably benchmark it using `lambkin`, but otherwise I think this is complete and ready for a review.

https://user-images.githubusercontent.com/13500507/148704332-46c77938-f82d-4574-875c-093ec19c8117.mp4

Vertices are removed yet the pose graph remains connected and sparse.
The current scoring mechanism results in large swathes of the map being forgotten rather quickly. I hope "our" information theoretic scoring does a better job.

nahueespinosa commented on slam_toolbox/lib/karto_sdk/include/karto_sdk/Mapper.h:
```diff
@@ -294,6 +312,19 @@ namespace karto
+        std::cout << "Edge not connected to vertex!" << std::endl;
+        return;
+      }
+      RemoveEdge(std::distance(m_Edges.begin(), it));
```
@hidmic nit: Is there a reason not to use `m_Edges.erase(it)`?

hidmic replied:
@nahueespinosa only to reuse `RemoveEdge(int)` functionality. If it were to change, we'd be covered and it's quite likely the compiler is inlining it anyways.

nahueespinosa commented on slam_toolbox/lib/karto_sdk/include/karto_sdk/Mapper.h:
```diff
@@ -632,6 +663,19 @@ namespace karto
+        std::cout << "Edge not found in graph!" << std::endl;
+        return;
+      }
+      RemoveEdge(std::distance(m_Edges.begin(), it));
```
@hidmic nit: same here, `m_Edges.erase(it)`?

hidmic replied:
@nahueespinosa same

nahueespinosa commented on slam_toolbox/lib/karto_sdk/include/karto_sdk/contrib/EigenExtensions.h:
```diff
@@ -0,0 +1,700 @@
+namespace internal {
+
+template<typename LhsType, typename RhsType>
+struct evaluator<HorizontalStack<LhsType, RhsType>>
```
@hidmic question: What are these evaluator classes used for?

hidmic replied:
@nahueespinosa generally speaking, these patterns are the backbone for Eigen's lazy evaluated, lightweight expressions (see [here](https://eigen.tuxfamily.org/dox/TopicNewExpressionType.html)). In this case, they implement most of the logic behind sparse matrix concatenation. 

I figured it was better than copying massive sparse matrices around multiples times. Also, caller code reads better :)

nahueespinosa replied:
@hidmic It's great! Thanks for the link.

nahueespinosa reviewed (commented):
This looks awesome!

hidmic committed:
Add node marginalization support.

Based on a Chow-Liu tree approximation to Karto's graph
upon marginalization, using Karto's solver estimate of
the full joint information matrix to build new nonlinear
constraints.

Signed-off-by: Michel Hidalgo <michel@ekumenlabs.com>

hidmic commented on slam_toolbox/lib/karto_sdk/include/karto_sdk/contrib/ChowLiuTreeApprox.h:
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2022 Ekumen Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
```
@agalbachicar is this appropriate for a copyright? License here is already LGPL, so I'm sticking to that.

agalbachicar replied:
I've checked with @patricioreyna and I confirm we can release it with this license note. I will reply tomorrow about the company name.

- [x] Resolve company name.

agalbachicar replied:
I could not find a project with `Ekumen` and our recent Noha simulation repo used BSD 3-Clause with `Ekumen Inc.`. Let's proceed with `Ekumen Inc.`.


About licensing and chats with @patricioreyna and @hidmic :

[GPL version 3](https://www.gnu.org/licenses/gpl-3.0.en.html), related clauses:

> 15. Disclaimer of Warranty.
THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM “AS IS” WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE PROGRAM IS WITH YOU. SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR OR CORRECTION.

> 16. Limitation of Liability.
IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MODIFIES AND/OR CONVEYS THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER PROGRAMS), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.

> 17. Interpretation of Sections 15 and 16.
If the disclaimer of warranty and limitation of liability provided above cannot be given local legal effect according to their terms, reviewing courts shall apply local law that most closely approximates an absolute waiver of all civil liability in connection with the Program, unless a warranty or assumption of liability accompanies a copy of the Program in return for a fee.

[LGPL version 3](http://www.gnu.org/licenses/lgpl-3.0.html#header) includes unrelated clauses atop GPL version 3.


All in all, you can proceed as is because Ekumen won't be hold liable nor warrant anything about this patch.

hidmic commented:
Alright, thank you both for the reviews and comments. I can't get Github to open a PR against upstream while this is open so I'll close it.
